src/Linters/UseCanonicalDataProviders.hack

```diff
@@ -0,0 +1,307 @@
+/*
+ *  Copyright (c) 2017-present, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under the MIT license found in the
+ *  LICENSE file in the root directory of this source tree.
+ *
+ */
+
+namespace Facebook\HHAST;
+
+use namespace HH\Lib\{C, Dict, Math, Str, Vec};
+
+final class UseCanonicalDataProvidersLinter extends AutoFixingASTLinter {
+  const type TContext = ClassishDeclaration;
+  const type TNode = FunctionDeclarationHeader;
+  // dict<provider, vec<(test, provider again)>>
+  const type TProviders = dict<string, vec<(string, string)>>;
+
+  const string T_DATAPROVIDER = 'Facebook\\HackTest\\DataProvider';
+  const string T_NOTHING = 'nothing';
+
+  <<__Override>>
+  public function getLintErrorForNode(
+    this::TContext $context,
+    this::TNode $node,
+  ): ?ASTLintError {
+    $info = $this->getInfoAboutTestFile($context);
+    if ($info is null) {
+      return null;
+    }
+
+    list($data_providers, $hhast_methods) = tuple(
+      $info['data_providers'],
+      $info['hhast_methods'],
+    );
+
+    $method_name = $node->getName()->getText();
+
+    if (!C\contains_key($data_providers, $method_name)) {
+      return null;
+    }
+
+    $hhast_method = $hhast_methods[$method_name];
+
+    list($err, $tuple_type) = $this->jugdeReturnTypes($hhast_method, $node);
+
+    if ($err is nonnull) {
+      return $err;
+    }
+
+    foreach ($data_providers[$method_name] as list($test, $_)) {
+      $err = $this->judgeTestTypes(
+        $node,
+        $tuple_type as nonnull,
+        $hhast_methods[$test],
+      );
+      if ($err is nonnull) {
+        return $err;
+      }
+    }
+
+    return null;
+  }
+
+  private function jugdeReturnTypes(
+    FunctionDeclarationHeader $hhast_method,
+    this::TNode $node,
+  ): (?ASTLintError, ?TupleTypeSpecifier) {
+    $return_type = $hhast_method->getType();
+
+    // Only in partial files
+    if ($return_type is null) {
+      return tuple(
+        new ASTLintError(
+          $this,
+          'Add a typehint to your dataprovider',
```

Comment:
`s/typehint/return type/`

Both more specific, and we don't really like the PHP 'typehint' term any more: it's not a hint in Hack